### PR TITLE
feat(steami_screen): Add steami_screen widget library.

### LINKED ISSUE #267: feat: Add steami_screen widget library.
## Description

Migrate `steami_screen.py` (665 lines) from micropython-steami-tutorials into a proper driver-style module.

## Source

`micropython-steami-tutorials/lib/steami_screen.py`

## Tasks

- [ ] Create `lib/steami_screen/manifest.py`
- [ ] Create `lib/steami_screen/steami_screen/__init__.py` (exports Screen + color constants)
- [ ] Create `lib/steami_screen/steami_screen/screen.py` (Screen class + all 10 widgets)
- [ ] Create `lib/steami_screen/steami_screen/colors.py` (from steami_colors.py — rgb_to_gray4, rgb_to_rgb565, rgb_to_rgb8)
- [ ] Adapt internal imports (`from .colors import *`)
- [ ] Add `lib/steami_screen/README.md`
- [ ] Verify `ruff check` passes

## Widget inventory (all must work after migration)

title, value, subtitle, bar, gauge, graph, menu, compass, watch, face

## Parent

- #266 — Migration umbrella issue

## Blocked by

Nothing — can start immediately

### LINKED ISSUE #271: feat: Freeze steami_screen and display wrappers in firmware.
## Description

Add steami_screen, steami_ssd1327, and steami_gc9a01 to the micropython-steami firmware manifest so they are frozen in ROM.

## Tasks

- [ ] Update micropython-steami manifest.py to include steami_screen
- [ ] Update micropython-steami manifest.py to include steami_ssd1327
- [ ] Update micropython-steami manifest.py to include steami_gc9a01 (if GC9A01 driver is available)
- [ ] `make firmware && make deploy`
- [ ] Verify `import steami_screen` works from REPL
- [ ] Verify tutorials work with frozen modules

## Dependencies

- #267 — steami_screen module
- #268 — steami_ssd1327 wrapper
- #269 — steami_gc9a01 wrapper

## Parent

- #266 — Migration umbrella issue

### PR DESCRIPTION
**Pull request by Charly-sketch**
## Summary

Add the new `steami_screen` widget library by migrating `steami_screen.py` from `micropython-steami-tutorials` into a driver-style module layout. 

Closes #267.
Closes #271.

Opens #348.

## Changes

- Add `lib/steami_screen/manifest.py`
- Add `lib/steami_screen/steami_screen/__init__.py` to export `Screen` and color helpers/constants
- Add `lib/steami_screen/steami_screen/screen.py` with the `Screen` class and all 10 widgets
- Add `lib/steami_screen/steami_screen/colors.py` migrated from `steami_colors.py`
- Adapt internal imports to use relative imports (`from .colors import *`)
- Migrate and preserve support for all existing widgets: `title`, `value`, `subtitle`, `bar`, `gauge`, `graph`, `menu`, `compass`, `watch`, `face`
- Run `ruff check` to validate the migrated module structure and code style
- Firmware integration:
  - Add all modules to firmware `manifest.py` for freezing
  - Verify modules are available at runtime without external `lib/`
  - When running make deploy, the high-level driver files (specifically steami_screen) are not updated on the board. However, make firmware works correctly: if the generated firmware.hex is manually copied to the board, the updated driver files are properly applied.

 
## Checklist

- [x] `ruff check` passes
- [ ] `python -m pytest tests/ -k mock -v` passes (no mock test broken)
- [ ] Tested on hardware (if applicable)
- [ ] README updated (if adding/changing public API)
- [ ] Examples added/updated (if applicable)
- [ ] Commit messages follow `<scope>: <Description.>` format

### CODE REVIEW

**Review by nedseb — COMMENTED** [review]
## Review & corrections applied

### What was reviewed

Migration de steami_screen depuis le repo tutorials. Code riche (659 lignes, 10 widgets), architecture propre (colors.py + device.py). Certains points de la review #346 ont été corrigés (width/height auto, gauge div/0, unused vars).

### Corrections applied

| Commit | Fix |
|---|---|
| `ff9be7c` | **Import cassé** : `from colors import rgb_to_gray4` → `from steami_screen.colors import rgb_to_gray4`. L'import implicite ne fonctionne pas en MicroPython frozen. |
| `52f0e9a` | **README** : ajout d'un README complet avec API reference, exemples de code, couleurs, utilitaires. |
| `257ba78` | **Tests mock** : 28 tests couvrant propriétés, primitives, texte, widgets (bar, menu, face, gauge, compass, watch, graph, circle), et edge cases. **Renommage** `screen.py` → `device.py` pour suivre la convention du projet (le test runner attend `{module}.device`). |

### Not changed

- **Exemples** : pas ajoutés dans cette PR (attendus dans une PR suivante de Charly)
- **Text scaling** : le scale=2 fait un effet bold, pas un vrai zoom — acceptable pour MicroPython, à documenter éventuellement

### Status

28 mock tests pass. Lint OK. Ready to merge (squash recommended).

**Comment by nedseb** [comment]
## Additional corrections (following your review)

Two real bugs found and fixed:

### 1. RGB tuples incompatible with SSD1327 backend (`0d6fbce`)

`Screen` was sending RGB tuples `(255, 255, 255)` to the backend, but the SSD1327 driver expects greyscale ints (0-15). Every call to `clear()`, `text()`, `pixel()`, `line()`, `rect(fill=True)` would raise `TypeError` on the real OLED.

**Fix:** Added `_c(color)` conversion method + `_text()` and `_pixel()` wrappers. All backend calls now go through these wrappers which auto-detect the backend type and convert RGB→grey4 when needed. The detection uses `_needs_gray = not hasattr(display, 'fill_rect') or hasattr(display, 'framebuf')`.

Also fixed a recursive call bug introduced during the refactor (`_text` calling itself instead of `self._d.text`).

### 2. `bar()` ZeroDivisionError when `max_val=0` (`0d6fbce`)

```python
fill_w = int(bar_w * min(val, max_val) / max_val)  # crashes if max_val=0
```

Fixed with `if max_val else 0`, consistent with the guard already present in `gauge()`.

### Updated correction summary

| Commit | Fix |
|---|---|
| `ff9be7c` | Import `from colors` → `from steami_screen.colors` |
| `52f0e9a` | README with full API reference |
| `257ba78` | 28 mock tests + rename `screen.py` → `device.py` |
| `169684b` | Document text scaling limitation |
| `0d6fbce` | **SSD1327 color conversion + bar() div/0** |

PR title also corrected to `feat(steami_screen): Add steami_screen widget library.`

### Status

28 mock tests pass. Lint OK. Ready to merge (squash recommended).

File: lib/steami_screen/steami_screen/device.py [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+                    self._fill_rect(ox + col * scale, oy + row * scale,
+                                    scale, scale, color)
+
+    # --- Level 2: Cardinal text & shapes ---
+
+    def text(self, text, at="CENTER", color=WHITE, scale=1):
+        """Draw text at a cardinal position or explicit (x,y)."""
+        if isinstance(at, str):
+            x, y = self._resolve(at, len(text), scale)
+        else:
```
[the remark this answers is not in the document]
Comment (nedseb) [reply lib/steami_screen/steami_screen/device.py:512]: Fixed in 0d6fbce: all backend calls now go through `_c()` / `_text()` / `_pixel()` wrappers that auto-convert RGB tuples to grey4 for SSD1327.

File: lib/steami_screen/steami_screen/device.py [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+            ly = yp - self.CHAR_H // 2
+            draw_sm(label, lx, ly, DARK)
+
+        # Dashed grid line at midpoint
+        mid_y = gy + gh // 2
+        dash, gap = 3, 3
+        x = gx + 1
+        while x < gx + gw:
+            x2 = min(x + dash - 1, gx + gw - 1)
+            self._d.line(x, mid_y, x2, mid_y, (51, 51, 51))
```
[the remark this answers is not in the document]
Comment (nedseb) [reply lib/steami_screen/steami_screen/device.py]: Fixed in 0d6fbce: grid line now uses `self._line()` which goes through color conversion.

File: lib/steami_screen/README.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+```
+
+Displays a large centered value with optional label above and unit below.
+
+---
+
+#### Progress Bar
+
+```python
+screen.bar(value=75, max_value=100)
```
[the remark this answers is not in the document]
Comment (nedseb) [reply lib/steami_screen/README.md]: Fixed in 7b86403: `bar(value=75, max_value=100)` → `bar(75, max_val=100)` to match the actual API.

File: lib/steami_screen/README.md [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+```
+
+---
+
+#### Gauge
+
+```python
+screen.gauge(value=60, min_value=0, max_value=100, unit="C")
```
[the remark this answers is not in the document]
Comment (nedseb) [reply lib/steami_screen/README.md]: Fixed in 7b86403: `gauge(value=60, min_value=0, max_value=100)` → `gauge(60, min_val=0, max_val=100)` to match the actual API.
